halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
+                    column_index: query.column_index,
+                    rotation: query.rotation,
+                    phase: query.phase,
+                })
+            }
+            ExpressionMid::Instance(query) => {
+                let (col, rot) = (Column::new(query.column_index, Instance), query.rotation);
+                let index = self.add_instance(col, rot);
+                Expression::Instance(InstanceQuery {
+                    index: Some(index),
+                    column_index: query.column_index,
+                    rotation: query.rotation,
+                })
+            }
+            ExpressionMid::Challenge(c) => Expression::Challenge(*c),
+            ExpressionMid::Negated(e) => Expression::Negated(Box::new(self.as_expression(e))),
+            ExpressionMid::Sum(lhs, rhs) => Expression::Sum(
+                Box::new(self.as_expression(lhs)),
+                Box::new(self.as_expression(rhs)),
+            ),
+            ExpressionMid::Product(lhs, rhs) => Expression::Product(
+                Box::new(self.as_expression(lhs)),
+                Box::new(self.as_expression(rhs)),
+            ),
+            ExpressionMid::Scaled(e, c) => Expression::Scaled(Box::new(self.as_expression(e)), *c),
+        }
+    }
+}
+
+/// This is a description of the circuit environment, such as the gate, column and
+/// permutation arrangements.
+#[derive(Debug, Clone)]
+pub struct ConstraintSystemV2Backend<F: Field> {
+    pub(crate) num_fixed_columns: usize,
+    pub(crate) num_advice_columns: usize,
+    pub(crate) num_instance_columns: usize,
+    pub(crate) num_challenges: usize,
+
+    /// Contains the index of each advice column that is left unblinded.
+    pub(crate) unblinded_advice_columns: Vec<usize>,
+
+    /// Contains the phase for each advice column. Should have same length as num_advice_columns.
+    pub(crate) advice_column_phase: Vec<u8>,
+    /// Contains the phase for each challenge. Should have same length as num_challenges.
+    pub(crate) challenge_phase: Vec<u8>,
+
+    pub(crate) gates: Vec<GateV2Backend<F>>,
+
+    // Permutation argument for performing equality constraints
+    pub(crate) permutation: permutation::Argument,
+
+    // Vector of lookup arguments, where each corresponds to a sequence of
+    // input expressions and a sequence of table expressions involved in the lookup.
+    pub(crate) lookups: Vec<lookup::ArgumentV2<F>>,
+
+    // Vector of shuffle arguments, where each corresponds to a sequence of
+    // input expressions and a sequence of shuffle expressions involved in the shuffle.
+    pub(crate) shuffles: Vec<shuffle::ArgumentV2<F>>,
+
+    // List of indexes of Fixed columns which are associated to a circuit-general Column tied to their annotation.
+    pub(crate) general_column_annotations: HashMap<metadata::Column, String>,
+}
+
+/// Witness calculator.  Frontend function
+#[derive(Debug)]
+pub struct WitnessCalculator<'a, F: Field, ConcreteCircuit: Circuit<F>> {
```

Comment:
Hve you tried with Lifetime Elision to see if we can get ridd of the `'a`s?? They're annoying as hell in the top levels of the user lib.

Reply:
https://doc.rust-lang.org/reference/lifetime-elision.html only talks about Lifetime Elision in functions and trait objects, so I'm not sure if anything can be elided here.

Reply:
Ahhhh that's true!!! My bad.

Well I guess we need to live with the `'a`. :smile: